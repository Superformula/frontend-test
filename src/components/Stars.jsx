import React from 'react';

const Stars = props => <div>{ props.numStars }</div>;

export default Stars;