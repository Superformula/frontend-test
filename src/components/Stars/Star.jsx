import React from 'react';
import { number } from 'prop-types';
import cx from 'classnames';

export const Star = ({ score }) => (
  <figure
    className={cx('star', {
      'star--filled': score > 0.5,
      'star--half': score === 0.5,
      'star--empty': score < 0.5,
    })}
  />
);

Star.propTypes = {
  score: number,
};

Star.defaultProps = {
  score: 0,
};

export default Star;
