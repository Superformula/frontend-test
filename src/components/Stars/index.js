import React, { PureComponent } from 'react';
import uniqueId from 'lodash/uniqueId';
import Icon from 'components/Icon';
import style from './stars.css';

const Stars = ({ number }) => {
  let remaining = 5;
  const stars = []
  const fullStars = Math.floor(number);

  // Add Full Stars
  if (number > 0) {
    for (let i = 0; i < number - 1; i++) {
      remaining--
      stars.push(<Icon name="fullStar" id={`fullStar_${uniqueId()}`} />)
    }
  }

  // Add Half Star
  if (number % 1 !== 0) {
    remaining--
    stars.push(<Icon name="halfStar" id={`halfStar_${uniqueId()}`} />);
  }

  // Add Empty Stars
  for (let i = 0; i < remaining; i++) {
    stars.push(<Icon name="emptyStar" id={`emptyStar_${uniqueId()}`} />)
  }

  return (
    <div className={style.container}>
      {stars}
    </div>
  )
}

export default Stars;
