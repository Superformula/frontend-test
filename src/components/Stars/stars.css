.container {
  padding-bottom: var(--medium-padding)
}
