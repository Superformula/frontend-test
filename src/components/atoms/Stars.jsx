import React from 'react';
import ReactStars from 'react-stars';

const Stars = props => {
    return (
        <ReactStars
            count={ 5 }
            edit={ false }
            value={ props.numStars }
            size={ 24 }
            color1={ 'grey' }
            color2={ '#021E44' }
        />
    );
}

export default Stars;