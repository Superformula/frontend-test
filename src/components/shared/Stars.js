import React, { memo } from 'react'
import PropTypes from 'prop-types'

function getStar (rating, i) {
  if (rating - i === 0.5) {
    return '◐'
  }

  return rating - i >= 1 ? '●' : '○'
}

function Stars ({ rating }) {
  let str = ''

  for (let i = 0; i < 5; i++) {
    str += getStar(rating, i)
  }

  return (
    <div className='stars' data-testid='stars'>
      {str}
    </div>
  )
}

Stars.propTypes = {
  rating: PropTypes.number
}

export default memo(Stars)
