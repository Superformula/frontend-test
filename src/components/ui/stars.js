import styled from "styled-components";
import PropTypes from "prop-types";
import React from "react";

import _Star from "./star";

const Container = styled.div`
  line-height: 1;
`;

const Star = styled(_Star)`
  margin-right: 2px;
`;

export function toArray(rating, max) {
  const arr = new Array(max);
  for (let i = 0; i < max; i++) {
    arr[i] = 0;
  }
  for (let i = 0; i < rating; i++) {
    arr[i] = 1;
  }
  const last = rating - Math.floor(rating);
  if (last > 0) {
    arr[Math.floor(rating)] = last;
  }
  return arr;
}

export default function Stars({ rating, max }) {
  // convert the rating to an array of numbers
  const arr = toArray(rating, max);
  return (
    <Container>
      {arr.map((rat, i) => (
        <Star key={i} rating={rat} />
      ))}
    </Container>
  );
}

Stars.propTypes = {
  rating: PropTypes.number.isRequired,
  max: PropTypes.number.isRequired,
};
