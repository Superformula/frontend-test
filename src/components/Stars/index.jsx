import React, { Component } from 'react';
import { number } from 'prop-types';
import { Star } from './Star';

export class StarsContainer extends Component {
  score = (amount) => {
    const { rating } = this.props;
    if (rating > amount && rating !== amount) return 1;
    if (rating < amount && rating > amount - 1) return 0.5;
    if (rating < amount) return 0;

    return 1;
  };

  renderStars() {
    const items = [];

    /* eslint-disable-next-line */
    for (let i = 1; i < 6; i++) {
      items.push(<Star key={i} score={this.score(i)} />);
    }

    return items;
  }

  render() {
    return <div className="stars">{this.renderStars()}</div>;
  }
}

StarsContainer.propTypes = {
  rating: number,
};

export default StarsContainer;
