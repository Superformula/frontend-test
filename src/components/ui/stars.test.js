import React from "react";
import Stars from "./stars";
import renderer from "react-test-renderer";
import { toArray } from "./stars";

test(`toArray converts 0 rating to [0,0,0,0]`, () => {
  const result = toArray(0, 4);
  expect(result).toEqual([0, 0, 0, 0]);
});

test(`toArray converts 0.5 rating to [0.5,0,0,0]`, () => {
  const result = toArray(0.5, 4);
  expect(result).toEqual([0.5, 0, 0, 0]);
});

test(`toArray converts 1 rating to [1,0,0,0]`, () => {
  const result = toArray(1, 4);
  expect(result).toEqual([1, 0, 0, 0]);
});

test(`toArray converts 1.5 rating to [1,0.5,0,0]`, () => {
  const result = toArray(1.5, 4);
  expect(result).toEqual([1, 0.5, 0, 0]);
});

test(`toArray converts 3 rating to [1,1,1,0]`, () => {
  const result = toArray(3, 4);
  expect(result).toEqual([1, 1, 1, 0]);
});

test(`toArray converts 3.5 rating to [1,1,1,0.5]`, () => {
  const result = toArray(3.5, 4);
  expect(result).toEqual([1, 1, 1, 0.5]);
});

test("Stars renders empty stars correctly", () => {
  const component = renderer.create(<Stars rating={0} max={0} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test("Stars renders correctly", () => {
  const component = renderer.create(<Stars rating={0} max={5} />);
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
