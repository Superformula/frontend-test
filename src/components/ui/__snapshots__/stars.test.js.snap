// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Stars renders correctly 1`] = `
<div
  className="stars__Container-m5uppc-0 kUWcna"
>
  <div
    className="star__Star-sc-11o45p7-0 stars__Star-m5uppc-1 fcJOLy"
  />
  <div
    className="star__Star-sc-11o45p7-0 stars__Star-m5uppc-1 fcJOLy"
  />
  <div
    className="star__Star-sc-11o45p7-0 stars__Star-m5uppc-1 fcJOLy"
  />
  <div
    className="star__Star-sc-11o45p7-0 stars__Star-m5uppc-1 fcJOLy"
  />
  <div
    className="star__Star-sc-11o45p7-0 stars__Star-m5uppc-1 fcJOLy"
  />
</div>
`;

exports[`Stars renders empty stars correctly 1`] = `
<div
  className="stars__Container-m5uppc-0 kUWcna"
/>
`;
